### PR TITLE
Update DIETClassifier to support composite entities

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
Update the `DIETClassifier` to support composite entities.

We are using multiple CRFs for that purpose. We first have a CRF for the entity label, followed by a CRF for the role label and one for the group label. The logits from the CRF for the entity label are passed onto the other CRFs. BILOU tagging is applied to the entity label as before. 

Documentation will be added in a separate PR.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: data/test/demo-rasa-composite-entities.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+- What's your name?
+- What can I call you?
+- How's the weather?
+- Is it too hot outside?
+
+## intent:restaurant_search
+- i'm looking for a place to eat
+- I want to grab lunch
+- I am searching for a dinner spot
+- i'm looking for a place in the [north](location) of town
```
Comment (Ghostvv) [inline data/test/demo-rasa-composite-entities.md:57]: is it equivalent to `i'm looking for a place in the [north]{"entity": "location"} of town`
I'm afraid different type of parenthesis will be confusing for users

File: rasa/nlu/classifiers/diet_classifier.py [hunk 2]
```diff
@@ -289,11 +298,12 @@ def __init__(
...
         # transform numbers to labels
         self.index_label_id_mapping = index_label_id_mapping
-        self.index_tag_id_mapping = index_tag_id_mapping
+        self.index_tag_id_mappings = index_tag_id_mappings
 
         self.model = model
 
-        self.num_tags: Optional[int] = None  # number of entity tags
+        # number of entity tags for every CRF
+        self.crf_num_tags: Optional[Dict[Text, int]] = None
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:306]: should this be private?
```suggestion
        self._crf_num_tags: Optional[Dict[Text, int]] = None
```

File: rasa/nlu/classifiers/diet_classifier.py [hunk 3]
```diff
@@ -289,11 +298,12 @@ def __init__(
...
         self._label_data: Optional[RasaModelData] = None
         self.data_example: Optional[Dict[Text, List[np.ndarray]]] = None
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:308]: maybe private?
```suggestion
        self._data_example: Optional[Dict[Text, List[np.ndarray]]] = None
```

File: rasa/nlu/classifiers/diet_classifier.py [hunk 4]
```diff
@@ -92,6 +99,8 @@
 LABEL_IDS = f"{LABEL}_ids"
 TAG_IDS = "tag_ids"
 
+CRF_ORDER = [ENTITY_ATTRIBUTE_TYPE, ENTITY_ATTRIBUTE_ROLE, ENTITY_ATTRIBUTE_GROUP]
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:102]: should this rather be `self._crf_order` property?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 5]
```diff
@@ -323,20 +333,50 @@ def _label_id_index_mapping(
...
+        tag_id_index_mapping_dict = {}
 
-        distinct_tag_ids = {
-            e["entity"]
-            for example in training_data.entity_examples
-            for e in example.get(ENTITIES)
-        } - {None}
+        for attribute_key in CRF_ORDER:
+            if (
+                attribute_key == ENTITY_ATTRIBUTE_TYPE
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:345]: should `BILOU` be added for all tags?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 6]
```diff
@@ -323,20 +333,50 @@ def _label_id_index_mapping(
...
+                )
+
+            if tag_id_index_mapping:
+                tag_id_index_mapping_dict[attribute_key] = tag_id_index_mapping
+
+        return tag_id_index_mapping_dict
+
+    @staticmethod
+    def _tag_id_index_mapping_for(
+        attribute_key: Text, entities: List[Message]
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:362]:
```suggestion
        attribute_key: Text, entity_examples: List[Message]
```

File: rasa/nlu/classifiers/diet_classifier.py [hunk 7]
```diff
@@ -323,20 +333,50 @@ def _label_id_index_mapping(
...
-    def _tag_id_index_mapping(self, training_data: TrainingData) -> Dict[Text, int]:
-        """Create tag_id dictionary"""
+    def _tag_id_index_mappings(
+        self, training_data: TrainingData
+    ) -> Dict[Text, Dict[Text, int]]:
+        """Create tag_id dictionaries for all CRFs."""
 
-        if self.component_config[BILOU_FLAG]:
-            return bilou_utils.build_tag_id_dict(training_data)
+        tag_id_index_mapping_dict = {}
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:341]: this names become ridiculous 🙈 is it a mapping to a mapping? Let's come up with a better name

File: rasa/nlu/classifiers/diet_classifier.py [hunk 8]
```diff
@@ -526,23 +566,15 @@ def _create_model_data(
...
-                    _tags = bilou_utils.tags_to_ids(e, tag_id_dict)
-                else:
-                    _tags = []
-                    for t in e.get(TOKENS_NAMES[TEXT]):
-                        _tag = determine_token_labels(t, e.get(ENTITIES), None)
-                        _tags.append(tag_id_dict[_tag])
-                # transpose to have seq_len x 1
-                tag_ids.append(np.array([_tags]).T)
+            if self.component_config.get(ENTITY_RECOGNITION) and tag_id_mappings:
+                self._update_crf_tag_ids(e, tag_id_mappings, crf_tag_ids)
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:570]: should mutable inputs be avoided?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 9]
```diff
@@ -556,13 +588,35 @@ def _create_model_data(
...
-        model_data.add_features(TAG_IDS, [tag_ids])
+        for name, tag_ids in crf_tag_ids.items():
+            model_data.add_features(f"{name}_{TAG_IDS}", [tag_ids])
 
         model_data.add_mask(TEXT_MASK, TEXT_FEATURES)
         model_data.add_mask(LABEL_MASK, LABEL_FEATURES)
 
         return model_data
 
+    def _update_crf_tag_ids(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:599]: I feel it has a lot in common with `_tag_id_index_mappings` can they be united?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 10]
```diff
@@ -727,16 +786,16 @@ def _predict_entities(
...
-
-        if self.component_config[BILOU_FLAG]:
-            tags = bilou_utils.remove_bilou_prefixes(tags)
+        predicted_tags = {}
+        for name, mapping in self.index_tag_id_mappings.items():
+            predictions = predict_out[f"e_{name}_ids"].numpy()
+            tags = [mapping[p] for p in predictions[0]]
+            if name == ENTITY_ATTRIBUTE_TYPE and self.component_config[BILOU_FLAG]:
+                tags = bilou_utils.remove_bilou_prefixes(tags)
+            predicted_tags[name] = tags
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:795]: I feel like, it deserves separate helper

File: rasa/nlu/classifiers/diet_classifier.py [hunk 11]
```diff
@@ -745,38 +804,74 @@ def _predict_entities(
 
         return entities
 
-    @staticmethod
     def _convert_tags_to_entities(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:807]: it became way to long. Can it be broken into smaller helpers?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 12]
```diff
@@ -945,7 +1041,7 @@ def __init__(
         self,
         data_signature: Dict[Text, List[FeatureSignature]],
         label_data: RasaModelData,
-        index_tag_id_mapping: Optional[Dict[int, Text]],
+        index_tag_id_mappings: Optional[Dict[Text, Dict[int, Text]]],
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1044]: do we need it only to calculate number of tags? Leet's pass `crf_num_tags` directly

File: rasa/nlu/classifiers/diet_classifier.py [hunk 13]
```diff
@@ -1018,19 +1120,24 @@ def _create_metrics(self) -> None:
         # so create loss metrics first to output losses first
         self.mask_loss = tf.keras.metrics.Mean(name="m_loss")
         self.intent_loss = tf.keras.metrics.Mean(name="i_loss")
-        self.entity_loss = tf.keras.metrics.Mean(name="e_loss")
+        self.entity_group_loss = tf.keras.metrics.Mean(name="group_loss")
+        self.entity_role_loss = tf.keras.metrics.Mean(name="role_loss")
+        self.entity_loss = tf.keras.metrics.Mean(name="entity_loss")
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1125]: do we want to output them in this order?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 14]
```diff
@@ -1274,7 +1380,7 @@ def _last_token(x: tf.Tensor, sequence_lengths: tf.Tensor) -> tf.Tensor:
         return tf.gather_nd(x, indices)
 
     def _f1_score_from_ids(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1382]: let's move this helper inside our `CRF` layer

File: rasa/nlu/classifiers/diet_classifier.py [hunk 15]
```diff
@@ -1447,10 +1594,23 @@ def batch_predict(
...
+            for name in CRF_ORDER:
+                if name not in self._crf_num_tags or self._crf_num_tags[name] == 0:
+                    continue
+
+                _input = text_transformed
+                if prev_logits is not None:
+                    _input = tf.concat([_input, prev_logits], axis=-1)
+
+                _logits = self._tf_layers[f"embed.{name}.logits"](_input)
+                pred_ids = self._tf_layers[f"crf.{name}"](_logits, sequence_lengths - 1)
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1607]: I think this part can be united with similar part in `_calculate_entity_loss` into a separate helper

File: data/test/demo-rasa-composite-entities.md (hunk 1 above)
Comment (tabergma) [reply data/test/demo-rasa-composite-entities.md:57]: We support both. In a standard case the user just needs `[north](location)` which is more readable. We discussed that and agreed that we are going to keep this simple format.

Comment (Ghostvv) [reply data/test/demo-rasa-composite-entities.md:57]: I meant shouldn't we use `(` instead of `{`?

Comment (tabergma) [reply data/test/demo-rasa-composite-entities.md:57]: Not sure if that is more intuitive as `{}` is used for a dict in python, which we are basically using here. But open for a discussion. However, I would not change the data format in this PR, if we decide to change it again, would create a new PR.

File: rasa/nlu/classifiers/diet_classifier.py (hunk 5 above)
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py]: Not sure, I can quickly run an experiment, but my assumption is that it does not change much. We also forward the logits from the first CRF, that includes BILOU, to the second level CRFs.

**Comment by tabergma** [comment]
Create a NamedTuple called `CRFLayer` to simplify the storage of the different mappings, tags for the individual CRFs. Let me know if you have a better name 😄

File: rasa/nlu/classifiers/diet_classifier.py (hunk 5 above)
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py]: Seems to be slightly better when we apply BILOU to all tags. Will update the code.

File: data/test/demo-rasa-composite-entities.md (hunk 1 above)
Comment (Ghostvv) [reply data/test/demo-rasa-composite-entities.md:57]: I just thought that users who create train data might not be familiar with any programming language. Any harm support both?

File: rasa/nlu/classifiers/diet_classifier.py [hunk 16]
```diff
@@ -1426,32 +1581,69 @@ def batch_predict(
...
-            logits = self._tf_layers["embed.logits"](text_transformed)
-            pred_ids = self._tf_layers["crf"](logits, sequence_lengths - 1)
-            out["e_ids"] = pred_ids
+            self._batch_predict_entities(out, sequence_lengths, text_transformed)
 
         return out
 
+    def _batch_predict_entities(
+        self,
+        out: Dict[Text, Any],
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1595]: I'd avoid having mutable input, I think it is better if this method returns `pred_ids`

File: rasa/nlu/classifiers/diet_classifier.py [hunk 17]
```diff
@@ -1426,32 +1581,69 @@ def batch_predict(
...
+            out[f"e_{name}_ids"] = pred_ids
+
+            if name == ENTITY_ATTRIBUTE_TYPE:
+                # use the logits from the entity type CRF as input for the role
+                # and group CRF
+                prev_logits = _logits
+
+    def _batch_predict_intents(
+        self,
+        out: Dict[Text, Any],
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1624]: same here, let's return instead

File: rasa/nlu/classifiers/diet_classifier.py [hunk 18]
```diff
@@ -1332,26 +1453,33 @@ def _calculate_label_loss(
...
+        inputs: tf.Tensor,
         tag_ids: tf.Tensor,
         mask: tf.Tensor,
         sequence_lengths: tf.Tensor,
-    ) -> Tuple[tf.Tensor, tf.Tensor]:
+        tag_name: Text,
+        previous_logits: Optional[tf.Tensor] = None,
+    ) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
 
         sequence_lengths = sequence_lengths - 1  # remove cls token
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1464]: this can be done before hand to avoid same operation be repeated in a for

File: rasa/nlu/classifiers/diet_classifier.py [hunk 19]
```diff
@@ -1426,32 +1581,69 @@ def batch_predict(
...
 
+    def _batch_predict_entities(
+        self,
+        out: Dict[Text, Any],
+        sequence_lengths: tf.Tensor,
+        text_transformed: tf.Tensor,
+    ) -> None:
+        prev_logits = None
+
+        for crf_layers in self._crf_layers:
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:1601]: in train method, this `for` is outside entity helper, let's unify this. Either outside in both train and predict helpers or inside in both

File: rasa/nlu/classifiers/diet_classifier.py [hunk 20]
```diff
@@ -556,13 +612,32 @@ def _create_model_data(
...
+        for tag_name, tag_ids in crf_tag_ids.items():
+            model_data.add_features(f"{tag_name}_{TAG_IDS}", [tag_ids])
 
         model_data.add_mask(TEXT_MASK, TEXT_FEATURES)
         model_data.add_mask(LABEL_MASK, LABEL_FEATURES)
 
         return model_data
 
+    def _tag_ids_for_crf(self, example: Message, crf_layer: CRFLayer) -> np.ndarray:
+        """Create a np.array containing the tag ids of the given message for the given
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:624]: If I remember correctly, first line of a doc string should be one sentence followed by a blank line, then additional description

File: rasa/nlu/classifiers/diet_classifier.py [hunk 21]
```diff
@@ -323,20 +348,53 @@ def _label_id_index_mapping(
     def _invert_mapping(mapping: Dict) -> Dict:
         return {value: key for key, value in mapping.items()}
 
-    def _tag_id_index_mapping(self, training_data: TrainingData) -> Dict[Text, int]:
-        """Create tag_id dictionary"""
+    def _create_crf_layers(self, training_data: TrainingData) -> List[CRFLayer]:
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:351]: same here I wouldn't call it create layers, we use `prepare_layers` in the model to create tf layers

File: rasa/nlu/classifiers/diet_classifier.py [hunk 22]
```diff
@@ -93,6 +100,15 @@
 TAG_IDS = "tag_ids"
 
 
+class CRFLayer(NamedTuple):
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py:103]: Yeah, do you have a good name in mind? Maybe `CRFSpecification`?

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:103]: it is crf independent though. So it's rather to do smth with `tags`

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:103]: maybe `TagSpec`?

File: data/test/demo-rasa-composite-entities.md (hunk 1 above)
Comment (tabergma) [reply data/test/demo-rasa-composite-entities.md:57]: Would not support both, might be confusing for the user. Let's move the discussion to Slack.

File: rasa/nlu/classifiers/diet_classifier.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py:103]: What about `EntityTag`?

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:103]: I think `Spec` or `Specification` or smth similar makes sense there, because just `EntityTag` suggests the name of a tag for some token

**Comment by tabergma** [comment]
@tmbo Do you have time to review the PR today? I would like to merge so that I can share the feature with our users on Monday.
